staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -306,24 +242,99 @@ func (s *EtcdOptions) ApplyWithStorageFactoryTo(factory serverstorage.StorageFac
 		return nil
 	}
 
-	if !s.complete {
-		return fmt.Errorf("EtcdOptions.Apply called without completion")
-	}
-
 	if !s.SkipHealthEndpoints {
 		if err := s.addEtcdHealthEndpoint(c); err != nil {
 			return err
 		}
 	}
 
-	if s.resourceTransformers != nil {
+	// setup encryption
+	if err := s.applyResourceTransformers(c); err != nil {
+		return err
+	}
+
+	c.RESTOptionsGetter = s.CreateRESTOptionsGetter(factory, c.ResourceTransformers)
+	return nil
+}
+
+func (s *EtcdOptions) CreateSimpleRESTOptionsGetter(resourceTransformers encryptionconfig.ResourceTransformers, tracker flowcontrolrequest.StorageObjectCountTracker) generic.RESTOptionsGetter {
+	storageConfig := s.StorageConfig
+	if storageConfig.StorageObjectCountTracker == nil {
+		storageConfig.StorageObjectCountTracker = tracker
+	}
+
+	return s.CreateRESTOptionsGetter(&SimpleStorageFactory{StorageConfig: s.StorageConfig}, resourceTransformers)
+}
+
+func (s *EtcdOptions) CreateRESTOptionsGetter(factory serverstorage.StorageFactory, resourceTransformers encryptionconfig.ResourceTransformers) generic.RESTOptionsGetter {
+	if resourceTransformers != nil {
 		factory = &transformerStorageFactory{
 			delegate:             factory,
-			resourceTransformers: s.resourceTransformers,
+			resourceTransformers: resourceTransformers,
 		}
 	}
+	return &StorageFactoryRestOptionsFactory{Options: *s, StorageFactory: factory}
+}
+
+func (s *EtcdOptions) applyResourceTransformers(c *server.Config) (err error) {
+	if c.ResourceTransformers != nil {
+		return nil
+	}
+	if len(s.EncryptionProviderConfigFilepath) == 0 {
+		return nil
+	}
+
+	ctxServer := wait.ContextForChannel(c.DrainedNotify())
+	ctxTransformers, closeTransformers := context.WithCancel(ctxServer)
+	defer func() {
+		// in case of error, we want to close partially initialized (if any) transformers
+		if err != nil {
+			closeTransformers()
+		}
+	}()
```

Comment:
I forget the rules for named returns, does this always run, even if the error is shadow assigned in an inner scope?

Reply:
it does